### PR TITLE
Training pipeline changes for CausalLM

### PR DESCRIPTION
**Pull request by pranjal-nntrainer**
- Add training data pipeline (TrainingDataGenerator) with tokenized text loading and next-token prediction labels
- Add training entry point (main_lora_train.cpp) with configurable learning rate, epochs, and sample limits
-  Add forwarding for training mode and calcDerivative/calcGradient implementation for backpropagation for all the layers required for training a CasualLM.
- For the above point, architectural changes need to happen. Will merge them once they are finished for all the relevant layers.

Signed-off-by: Pranjal Thapliyal [p.thapliyal@samsung.com](mailto:p.thapliyal@samsung.com)

### CODE REVIEW

File: Applications/CausalLM/models/transformer.cpp [hunk 1]
```diff
@@ -214,6 +265,14 @@ void Transformer::constructModel() {
              "layer" + std::to_string(NUM_LAYERS - 1) + "_decoder_output"),
      withKey("packed", "false")}));
 
+  // create lm_head
+  const std::string lm_head_type =
+    TIE_WORD_EMBEDDINGS ? "tie_word_embeddings" : "fully_connected";
+  layers.push_back(createLayer(
+    lm_head_type, {"name=lm_head", "unit=" + std::to_string(NUM_VOCAB),
```
[the remark this answers is not in the document]
Comment (EunjuYang) [reply Applications/CausalLM/models/transformer.cpp:272]: ### Duplicate lm_head in base Transformer::constructModel()

I agree with this comment. `Applications/CausalLM/models/transformer.cpp:268-275` unconditionally appends an lm_head layer to the base graph. However, the derived class (e.g. `CausalLM::constructModel()` in `Applications/CausalLM/models/causal_lm.cpp`, which already adds output_of_causallm consuming output_norm) append their own output head. (Please note that this transformer class is inherited by both SentenceTransformer and CausalLM; CausalLM only adopts lmhead for its generation). As a result, every subclass that already has a head will now end up with two heads in the graph.

https://github.com/nntrainer/nntrainer/blob/57f8bbf846ad2cc8f383d80a2c0fe99ebd8542bc/Applications/CausalLM/models/causal_lm.cpp#L104-L126

File: Applications/CausalLM/layers/embedding_layer.cpp [hunk 2]
```diff
@@ -174,26 +213,21 @@ void EmbeddingLayer::save(std::ofstream &file,
...
                           ml::train::ExecutionMode mode, bool trainable,
                           nntrainer::TensorDim::DataType dtype) const {
-  // @note shared weights are only be saved at the first access
   for (unsigned int i = 0; i < run_context.getNumWeights(); ++i) {
     if (run_context.isGradientFirstAccess(i)) {
       auto &weight = run_context.getWeight(i);
-      if (dtype == nntrainer::TensorDim::DataType::NONE ||
-          weight.getDataType() == dtype)
+      if (weight.getDataType() == dtype)
         weight.save(file);
```
[the remark this answers is not in the document]
Comment (EunjuYang) [reply Applications/CausalLM/layers/embedding_layer.cpp:220]: I agree with this comment. I think you should not remove this condition:
```
if(dtype == nntrainer::TensorDim::DataType::NONE || ...)
```

File: Applications/CausalLM/layers/tie_word_embedding.cpp [hunk 3]
```diff
@@ -305,11 +327,164 @@ void TieWordEmbedding::incremental_forwarding_lmhead(
...
+        b * in.getDim().getFeatureLen() + (seq_len - 1) * hidden_dim;
+      nntrainer::Tensor in_last =
+        in.getSharedDataTensor(last_pos_dim, last_pos_offset, true);
+
+      // dy^T @ in_last: [vocab_size, 1] × [1, hidden_dim] = [vocab_size,
+      // hidden_dim] Accumulate into dweight
+      nntrainer::Tensor dw_temp(dweight.getDim());
+      dy.dot(in_last, dw_temp, true, false);
+      dweight.add_i(dw_temp);
+    }
```
Comment (EunjuYang) [inline Applications/CausalLM/layers/tie_word_embedding.cpp:458]: ### Gradient uses full dy instead of the per-batch slice (incorrect for batch > 1)

In Applications/CausalLM/layers/tie_word_embedding.cpp inside `calcGradient()` (lm_head branch):

```
for (unsigned int b = 0; b < b_size; ++b) {
  nntrainer::Tensor in_last = in.getSharedDataTensor(
      last_pos_dim, last_pos_offset, true);

  nntrainer::Tensor dw_temp(dweight.getDim());
  dy.dot(in_last, dw_temp, true, false);   // ← uses full dy every iteration
  dweight.add_i(dw_temp);
}
```
`dy` has shape `[batch, 1, 1, vocab_size]` but only a single-batch slice `[1, 1, 1, vocab_size]` should be used per iteration to pair with in_last.

**Review by EunjuYang — CHANGES_REQUESTED** [review]
### Suggest splitting / re-basing this PR on top of the in-flight layer PRs

This PR currently overlaps significantly with three other open PRs, all of which (a) implement the backward pass for a specific CausalLM layer and (b) move that layer from `Applications/CausalLM/layers/` to `nntrainer/layers/`:

- #3859 — RMSNorm forward/backward + layer move. This PR also rewrites Applications/CausalLM/layers/rms_norm.cpp with its own backward implementation.
- #3863 — LmHead forward/backward + layer move. This PR also rewrites Applications/CausalLM/layers/lm_head.cpp.
- #3865 — ReshapedRMSNorm forward/backward + layer move. This PR also rewrites Applications/CausalLM/layers/reshaped_rms_norm.cpp.

Whichever PR lands first will force the others into a large rebase, and the fact that #3859/#3863/#3865 move the files in addition to changing them makes rebasing especially painful if #3864 lands first — it would turn into "move + reapply someone else's backward on top of ours". The PR description here already notes "architectural changes need to happen. Will merge them once they are finished for all the relevant layers.", which matches this concern.

**Could we:**

1. Mark this PR as draft (or prepend [Depends on #3859, #3863, #3865] to the title) and add a Dependency section to the body listing those three PRs with wait-for semantics.
2. Once #3859 / #3863 / #3865 merge, rebase and drop the `rms_norm.cpp` / `lm_head.cpp` / `reshaped_rms_norm.cpp` edits that duplicate their work.
3. Keep this PR scoped to what is uniquely new here — the training entry point (`main_lora_train.cpp`, `lora_train.{h,cpp}`), the `Transformer::initializeForTraining` + `LoRA` wiring, and the backward passes for the remaining layers (`Embedding`, `SwiGLU`, `MHACore`, `TieWordEmbedding`).

That should shrink the diff to roughly half its current size and let each layer's backward be reviewed once, by its owner, instead of being approved twice in two different places.

**Comment by baek2sm** [comment]
**Review verdict: 💬 Comment** (rebased and x86 core green, but the actual training code is exercised by no CI job and is unverifiable in our environment — not approving)

### 1. Summary
Adds a CausalLM training pipeline: dataset wiring + train loop on `Transformer` (`setDataset`/`train`/`initializeForTraining`), a LoRA training entry point (`lora_train.cpp`), LoRA props threaded into `createAttention`/`createMlp`, and a training-forward path in `MHACoreLayer`. The branch was rebased onto the latest `upstream/main` (authorship preserved; commit bodies and sign-off completed for DCO). The rebase adapted `initializeForTraining()` to the current symbolic-graph API (`constructModel() → {x,h}` then `compile(x, h, TRAIN)`), and follow-up commits fixed trailing newlines, clang-format on the training-forward block, and doxygen params.

### 2. CI status / why approval is withheld
The core library builds and all 25 x86 unit tests pass (Windows MSVC, locally). However, the training-pipeline code that this PR is *about* — `transformer.cpp`, `lora_train.cpp`, and the `MHACoreLayer` training path — is `enable-transformer`-gated and is **not compiled by any current CI job** (`causallm_android.yml` builds the inference targets, not `nntr_causallm_lora_train`). A green CI here therefore does **not** verify the training pipeline.

### 3. What needs consideration for verification (and our limitation)
We can only run the **Windows x86 MSVC** default build, which compiles **none** of the training code. If you verify, focus on:
- Build with `enable-transformer=true` (and add `lora_train.cpp` to the Android.mk/CI if you want CI coverage), then run the LoRA training end-to-end (`download_sst2.py` dataset) to confirm loss decreases and a checkpoint is produced.
- `MHACoreLayer` training-forward path: the raw `getAddress<float>()` accesses (lines ~449/471/506/522) assume FP32 — verify behavior if activation dtype is not FP32.
- `apply_rotary_emb_tensor_v2(q_rope, q_rope, ...)` is called with input == output (aliased) on the training path — confirm the kernel is safe under aliasing (the inference path does not alias).
- `initializeForTraining()` creates its own model handle + optimizer — confirm it isn't called alongside `initialize()` in a way that double-initializes state.

The rebase adaptation looks structurally correct and the x86 core is unaffected, but the training feature itself is unverified (no CI compiles it, and we cannot build it locally) — hence a comment rather than an approval.
